lib/public/views/lhcPeriods/format/formatDistinctLhcBeamEnergies.js

```diff
@@ -13,81 +14,50 @@
-            : lhcBeamEnergy.toFixed(notChangedEnergyPrecision).toString();
+function getClosestEnergyLevel(lhcBeamEnergy) {
+    for (const energyLevel of WELL_DEFINED_ENERGY_LEVELS) {
+        if (lhcBeamEnergy < energyLevel - ENERGY_LEVEL_MARGIN) {
+            return null;
+        } else if (lhcBeamEnergy <= energyLevel + ENERGY_LEVEL_MARGIN) {
+            return energyLevel;
+        }
+    }
+    return null;
+}
 
 /**
- * Takes lhcBeamEnergy and return it or closest defined energy accoriding to the following conditions.
+ * Display a given LHC beam energy
  *
- *      If consider all energies `E DEFINED IN wellDefinedEnergyLevels`
- *      and take minimal absolute value of difference of lhcBeamEnergy and E
- *      and `|lhcBeamEnergy - closestEnergy| <= acceptableMargin`
- *      then the function returns `closestEnergy`
- *      if there is no `closestEnergy` fulfilling all requirements then the function returns `lhcBeamEnergy`.
+ * If the energy is close enough (within a given margin)
+ * to a well defined energy level, displays the well defined energy level instead
  *
- * @param {number} lhcBeamEnergy list of energies to be transformed
- * @returns {string} closest energy in displayable format
+ * @param {number} lhcBeamEnergy the energy to display
+ * @return {string} the energy display
  */
-const distinctLhcBeamEnergyToDisplay = (lhcBeamEnergy) => {
-    const closestEnergy = getClosestDefinedEnergyLevel(
-        lhcBeamEnergy,
-        Object.keys(distinctEnergiesProcessingConfiguration.wellDefinedEnergyLevelToDisplay),
-    );
-    return _distinctLhcBeamEnergyToDisplay(
-        lhcBeamEnergy,
-        closestEnergy,
-        distinctEnergiesProcessingConfiguration.wellDefinedEnergyLevelToDisplay,
-        distinctEnergiesProcessingConfiguration.acceptableEnergyMargin,
-    );
+const displayLhcBeamEnergy = (lhcBeamEnergy) => {
+    const energyLevel = getClosestEnergyLevel(lhcBeamEnergy);
+    const energyToDisplay = energyLevel !== null ? energyLevel : lhcBeamEnergy;
+    return energyToDisplay in ENERGY_LEVELS_CUSTOM_NAME
+        ? ENERGY_LEVELS_CUSTOM_NAME[energyToDisplay]
+        : energyToDisplay.toFixed(DISPLAY_PRECISION);
 };
 
 /**
- * Format lhc beam distinct energies, optionaly replacing some by closest energies
- * @param {number[]} distinctEnergies list of energies to be transformed
- * @returns {Component} list of distinct energies
+ * Format lhc beam distinct energies, optionaly replacing some of them
+ * by well defined energy levels @see displayLhcBeamEnergy
+ * @param {number[]} distinctEnergies list of energies to be displayed
+ * @returns {Component} formatted list of distinct energies
  */
 export const formatDistinctLhcBeamEnergies = (distinctEnergies) =>
     distinctEnergies?.length > 0
         ? h('', distinctEnergies
-            ?.map(distinctLhcBeamEnergyToDisplay)
+            ?.map(displayLhcBeamEnergy)
             .filter((value, index, array) => array.indexOf(value) === index) // Unique
```

Comment:
I don't know which one is the more optimized, but I would rather create a `Set` and go back to the array

Reply:
/**
 * Reason for calculation of runs lhc beam distinct energies this way (using literals) is:
 * There is transitive association: `runs -> lhc_periods -> lhc_periods_statistics`.
 * Including runs in queries below causes that sequelize build internaly subquery to fetch data.
 * This subquery is an `SELECT ... FROM lhc_periods_statistics`,
 * but after adding ordering by `lhcPeriod.name` sequelize puts the `ORDER BY` clause inside that subquery,
 * what leads to error, because lhcPeriod is not in scope of the subquery.
 * Mentioned can be avoid by providing flag:  `subQuery: false`,
 * but this breaks limit clause, so pagination.
 * There is related issue for reference: https://github.com/sequelize/sequelize/issues/11983
 */

Reply:
same